### PR TITLE
feat: adding composable for employees

### PR DESCRIPTION
**Pull request by rafaelmils**
# Changes

I have create a 'useEmployees' composable which we can re-use in every vue component if we want to retrieve
- a list the employes
- a list of all employess as items for a table
- a function with which we can retrieve an employee by **id**
- a computed which returns the employee based on the employee id route param

## How to test

- You should be able to see the employees list with all the employees when navigating to the Employees page. 
- When navigating to the Timesheets page, you should be able to see the employees list in the left side bar as well as clicking to a specific box(check screenshot below) next to the employee name, should show the calendar of the employee's hours for current week in the right side of the page.

## Screenshots
Screenshot 2023-01-02 at 11 35 59

### CODE REVIEW

**Review by marioesuc — DISMISSED** [review]

**Comment by sanderdejong88** [comment]
Looks good! There is one more file which uses the employees from the store: `_week.vue`
Want to move that logic as well?

**Comment by rafaelmils** [comment]
@sanderdejong88 I will incorporate the changes also to the file you mentioned! `_week.vue`

File: composables/useEmployees.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import {checkEmployeeAvailability} from '~/helpers/employee';
+
+export function useEmployees() {
+  const store = useStore<RootStoreState>();
+  const route = useRoute();
+  const employees = computed(() => store.state.employees.employees);
+
+  const showEmployeeError = ref<Boolean>(false);
+  const inactive = ref<boolean>(false);
+  const billable = ref<boolean>(false);
```
Comment (jurgenbelien) [inline composables/useEmployees.ts:11]: I think these variable names are a bit confusing, IMHO it should be `includeInactive` and `includeNonBillable`

File: composables/useEmployees.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  const employeeTableItems = computed(() =>
+    employees.value
+      .map(employee => ({
+        ...employee,
+        active: checkEmployeeAvailability(employee, new Date()),
+      }))
+      .filter(
+        employee =>
+          (inactive.value ? true : employee.active) && (billable.value ? true : employee.billable)
+      )
```
Comment (jurgenbelien) [inline composables/useEmployees.ts:26]: Maybe wise to spin this out in a separate function, and ditch the ternary statements

```typescript
.filter(applyEmployeeFilters)
```

```typescript
function applyEmployeeFilters({ active: isActive, billable: isBillable }: Employee) {
  return (isActive || includeInactive.value) && (isBillable || includeNonBillable.value)
}
```

File: composables/useEmployees.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    if (!employee) {
+      showEmployeeError.value = true;
+      return;
+    }
+
+    showEmployeeError.value = false;
+
+    if (employee?.id === store.state.employee.employee?.id) return employee;
+
+    if (store.state.employee.employee?.isAdmin) return employee;
```
Comment (jurgenbelien) [inline composables/useEmployees.ts:40]: I know this is not your code, but to me this looks a bit weird. 

I would expect by design a non-admin user cannot get a list of all employees. Having logic in this composable that tries to enforce authorization is completely futile. I can easily mutate the state to allow myself admin rights, and if it is already in de state, I can dump it just as easily.

This function should either return an employee or throw the employee error, that's it.

File: composables/useEmployees.ts (hunk 1 above)
Comment (jurgenbelien) [reply composables/useEmployees.ts:11]: Additionally, could you explain why these are in the `useEmployees()` composable, instead of in the view?

**Review by jurgenbelien — COMMENTED** [review]
I have some qualms about the `employeeTableItems`. I feel that it does not belong in the composable. I can imagine you'd want a `getFilteredEmployees()`or something?

File: composables/useEmployees.ts (hunk 1 above)
Comment (rafaelmils) [reply composables/useEmployees.ts:11]: I consider both of your points valid ones!
I have applied the more descriptive names, `includeInactive`, `includeNonBillable`.

I thought that maybe it would be nice to be able to export the table items immediately from the composable, but it makes sense what you mention here, it doesn't look reasonable to be there, thus I have moved them in the component along with the computed that returns the filtered items.

File: composables/useEmployees.ts (hunk 2 above)
Comment (rafaelmils) [reply composables/useEmployees.ts:26]: @jurgenbelien Also valid ones here, so I have applied some changes based on your comments.

File: composables/useEmployees.ts (hunk 3 above)
Comment (rafaelmils) [reply composables/useEmployees.ts:25]: @jurgenbelien  I could agree with you in this one but I will need to check with @sanderdejong88 if we really need that and what would be the best place to have that logic.

Comment (sanderdejong88) [reply composables/useEmployees.ts:25]: @jurgenbelien I tried doing that via a firebase security rule a long time ago but couldn't get that to work. It's easy to block access to entire collections, but blocking access to all records except the logged-in employee's record was quite tricky.

I'm also fine with completely removing the check. The worst thing that could happen is that you can see the hours a colleague wrote. But you would need to know their UUID to access the URL.

Comment (rafaelmils) [reply composables/useEmployees.ts:25]: @sanderdejong88 All right, since it also looks fine by removing that check, I will remove it then!

**Review by marioesuc — DISMISSED** [review]

File: components/reports/date-reports.vue [hunk 4]
```diff
@@ -255,17 +256,12 @@ export default defineComponent({
...
       const id = props.employeeId;
 
-      const employee = employees.value.find(e => e.id === id);
-      if (!employee) return;
-
-      if (employee.id === store.state.employee.employee?.id) return employee;
-
-      if (store.state.employee.employee?.isAdmin) return employee;
+      return getEmployeeById(id);
     });
```
Comment (jurgenbelien) [inline components/reports/date-reports.vue:265]: You can write this even more succinct:

```typescript
const employee = computed(() => getEmployeeId(props.employeeId));
```

File: composables/useEmployees.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    showEmployeeError.value = false;
+
+    if (employee?.id === store.state.employee.employee?.id) return employee;
+  };
+
+  const employeeByRouteParamId = computed(() => {
+    return getEmployeeById(route.value.params.employee_id);
+  });
```
Comment (jurgenbelien) [inline composables/useEmployees.ts:28]: Couldn't you make this a oneliner now, it is not computed, right?

```typescript
const employeeByRouteParamId = computed(() => getEmployeeId(route.value.params.employee_id));
```

File: pages/admin/employees/index.vue [hunk 6]
```diff
@@ -39,10 +39,10 @@ nl:
             </b-input-group>
           </b-form-group>
 
-          <b-form-checkbox v-model="inactive" switch class="mr-4">
+          <b-form-checkbox v-model="includeInactive" switch class="mr-4">
```
Comment (jurgenbelien) [inline pages/admin/employees/index.vue:42]: Nice :)

File: pages/admin/employees/index.vue [hunk 7]
```diff
@@ -156,25 +157,26 @@ export default defineComponent({
...
-            (inactive.value ? true : employee.active) &&
-            (billable.value ? true : employee.billable)
-        )
+        .filter(applyEmployeeFilters)
     );
 
+    const applyEmployeeFilters = ({ active: isActive, billable: isBillable }: Employee & {active: boolean}) => {
+      return (isActive || includeInactive.value) && (isBillable || includeNonBillable.value)
+    }
+
```
Comment (jurgenbelien) [inline pages/admin/employees/index.vue:173]: Nice! But might be wise to move this above line 160, so that it is defined before being used

File: components/reports/date-reports.vue (hunk 4 above)
Comment (rafaelmils) [reply components/reports/date-reports.vue:265]: I totally agree! applied change!

File: pages/admin/employees/index.vue (hunk 7 above)
Comment (rafaelmils) [reply pages/admin/employees/index.vue:173]: Also valid one!

File: composables/useEmployees.ts [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    const employee = employees.value.find(e => e.id === id);
+    if (!employee) {
+      showEmployeeError.value = true;
+      return;
+    }
+
+    showEmployeeError.value = false;
+
+    if (employee?.id === store.state.employee.employee?.id) return employee;
+    if (store.state.employee.employee?.isAdmin) return employee;
```
Comment (rafaelmils) [inline composables/useEmployees.ts:24]: @sanderdejong88 I needed to revert the change of removing this line, as by removing it, I was not able to to get an employee back on the timesheets page. Apart from that, I have fixed all the remarks and we can proceed with merging.

File: pages/admin/employees/index.vue [hunk 9]
```diff
@@ -156,25 +157,26 @@ export default defineComponent({
       {key: 'billable', label: 'billable', sortable: false, class: 'text-center'},
       {key: 'actions', label: 'actions', sortable: false, class: 'text-right'},
     ];
-    const items = computed(() =>
+
+    const applyEmployeeFilters = ({ active: isActive, billable: isBillable }: Employee & {active: boolean}) => {
```
Comment (marioesuc) [inline pages/admin/employees/index.vue:161]: I'm not too much of a fan of this type union with `active`, although I understand it's difficult to get rid of in the scope of this PR because this key was already being added dynamically in the old code in the wrong place. What about making it optional in the `Employee` type directly?

Comment (rafaelmils) [reply pages/admin/employees/index.vue:161]: Valid one! Not a bad idea and definitely cleaner and more holistic that the existing one! :)

**Review by marioesuc — APPROVED** [review]

**Review by sanderdejong88 — APPROVED** [review]
